package:core XML parser to cover the case when the xml only contains the root level tag

ale-delfino opened the pull request:
Description: Fix xml parser to handle strings that only contain the root tag
Issue: N/A
Dependencies: None
Twitter handle: N/A

A valid xml text can contain only the root level tag. Example: <body>
  Some text here
</body>
The example above is a valid xml string. If parsed with the current implementation the result is {"body": []}. This fix checks if the root level text contains any non-whitespace character and if that's the case it returns {root.tag: root.text}. The result is that the above text is correctly parsed as {"body": "Some text here"}

@ale-delfino

Thank you for contributing to LangChain!

Checklist:

- [x] PR title: Please title your PR "package: description", where "package" is whichever of langchain, community, core, experimental, etc. is being modified. Use "docs: ..." for purely docs changes, "templates: ..." for template changes, "infra: ..." for CI changes.
  - Example: "community: add foobar LLM"
- [x] PR message: **Delete this entire template message** and replace it with the following bulleted list
    - **Description:** a description of the change
    - **Issue:** the issue # it fixes, if applicable
    - **Dependencies:** any dependencies required for this change
    - **Twitter handle:** if your PR gets announced, and you'd like a mention, we'll gladly shout you out!
- [x] Pass lint and test: Run `make format`, `make lint` and `make test` from the root of the package(s) you've modified to check that you're passing lint and testing. See contribution guidelines for more information on how to write/run tests, lint, etc: https://python.langchain.com/docs/contributing/
- [x] Add tests and docs: If you're adding a new integration, please include
  1. a test for the integration, preferably unit tests that do not rely on network access,
  2. an example notebook showing its use. It lives in `docs/docs/integrations` directory.

Additional guidelines:
- Make sure optional dependencies are imported within a function.
- Please do not add dependencies to pyproject.toml files (even optional ones) unless they are required for unit tests.
- Most PRs should not touch more than one package.
- Changes should be backwards compatible.
- If you are adding something to community, do not re-import it in langchain.

If no one reviews your PR within a few days, please @-mention one of @baskaryan, @efriis, @eyurtsev, @hwchase17.

ale-delfino committed:
fix(xml parser) XML parser to cover the case when the xml only contains the rrot level tag

A valid xml text can contain only the root level tag. Example:
<body>
  Some text here
</body>
The example above is a valid xml string. If parsed with the current implementation the result is {"body": []}.
This fix checks if the root level text contains any non-whitespace character and if that's the case it returns {root.tag: root.text}.
The result is that the above text is correctly parsed as {"body": "Some text here"}

@ale-delfino

hwchase17 reviewed (commented):
can we add a test case for this?

ale-delfino committed:
Adding a test for an xml text that only contains root level tag

ale-delfino committed:
Fixing the expected returned type

Before it was Dict[str, List[Any]] which works if there are more tags than only the root.

If there's only the root level tag we expect to return the content of the root as text so the returned type is Dict[str, str]

To cover for both cases Union[Dict[str, str], Dict[str, List[Any]]] is now returned

ale-delfino commented:
@hwchase17 I've added a simple test

ale-delfino committed:
Merge branch 'langchain-ai:master' into fix-xml-parser

eyurtsev committed:
Merge branch 'master' into fix-xml-parser

eyurtsev committed:
x

eyurtsev committed:
x

baskaryan committed:
Merge branch 'master' into fix-xml-parser

baskaryan committed:
fmt
